text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,460 @@
+
+OTLP is a request/response style protocols: the clients send requests, the server replies with corresponding responses. There are 2 supported requests and response types: `Hello` and `Export`.
+
+### Hello Handshake
+
+The message exchange between the client and server SHOULD start by the client initiating the underlying transport and sending `Hello` request. The server replies with `Hello` response. `Hello` request and response have no payload in this version of OTLP. They are intended to be extended in the future with a payload that describes client's and server's extended capabilities when such new capabilities are added to the protocol. The initial exchange of `Hello` messages is a handshake that allows client and server to discover each other's capabilities in a backwards compatible manner (see more about how future versions of the protocol will ensure interoperability in a separate section below).
+
+The client MAY choose to avoid sending `Hello` request and proceed to sending `Export` requests immediately after the transport is established (see the next section for `Export` description). From the server's perspective this behavior should be equivalent to a successful exchange of  `Hello` request and response with empty payloads.
+
+The server MUST always reply with a corresponding `Hello` response to every `Hello` request.
+
+### Export Request and Response
+
+After `Hello` request and response the client starts sending telemetry data using `Export` requests.  The client continuously sends a sequence of `Export` requests to the server and expects to receive a response to each request:
+
+![Request-Response](images/otlp-request-response.png)
+
+`Export` request contains telemetry data and includes an ID field that is unique for each request. The value of ID field is a sequential uint64 number generated by the client. `Export` response received from the server contains an ID field that acknowledges the receipt of the request with the same ID value.
+
+_Note: this protocol is concerned with reliability of delivery between one pair of client/server nodes and aims to ensure that no data is lost in-transit between the client and the server. Many telemetry collection systems have multiple nodes that the data must travel across until reaching the final destination (e.g. application -> agent -> collector -> backend). End-to-end delivery guarantees in such systems is outside of the scope of OTLP. The acknowledgements described in this protocol happen between a single client/server pair and do not span multiple nodes in multi-hop delivery paths._
+
+### Client Modes
+
+OTLP defines 2 modes of operation for clients: Synchronous and Pipelined.
```

Comment:
The Load Balancing discussion in this spec is targetted for the case when services behind Load Balancer do not require that all the requests go to the same instance. The goal is to allow Load Balancers rebalance traffic when needed.

So, you are right that unary gRPC would be enough from functional perspective, because unary calls can be rebalanced out of the box.

The problem with unary calls is bad performance. Benchmark results show about 1.12 times more CPU usage of unary calls compared to streaming implementation. This is one of the reasons I decided to go with gRPC streams.